File: elixir/lib/2022/16_proboscidea_volcanium/aoc202216.ex
defmodule AOC2022.Day16 do
  @moduledoc """
  Advent of Code 2022, day 16: Proboscidea Volcanium.
  """
  require AOC

  @doc """
  Parse input.
  """
  def parse(puzzle_input),
    do: puzzle_input |> String.split("\n", trim: true) |> Enum.into(%{}, &parse_valve/1)

  @doc """
  Parse information about one valve.

  ## Examples:

      iex> parse_valve("Valve BB has flow rate=13; tunnels lead to valves CC, AA")
      {"BB", {13, ["CC", "AA"]}}
      iex> parse_valve("Valve HH has flow rate=22; tunnel leads to valve GG")
      {"HH", {22, ["GG"]}}
  """
  def parse_valve(line) do
    [_, valve, rate, valves] =
      Regex.run(
        ~r/^Valve ([A-Z]{2}) has flow rate=(\d+); tunnel.+valves? (.+)$/,
        line
      )

    {valve,
     {rate |> String.to_integer(),
      valves |> String.split(",", trim: true) |> Enum.map(&String.trim/1)}}
  end

  @doc """
  Solve part 1.
  """
  def part1(valves), do: valves |> find_highest_flow("AA", 30) |> elem(1)

  @doc """
  Solve part 2.

  Currently too slow to be practical :(
  """
  def part2(valves),
    do: valves |> find_highest_flow_with_help("AA", 26) |> elem(1)

  @doc """
  Use Dijkstra to find the total risk of the safest path through the grid.

  """
  def find_highest_flow(valves, start, minutes) do
    total_flow = %{{start, 0, %MapSet{}} => 0}
    queue = Heap.max() |> Heap.push({0, 0, start, %MapSet{}})
    dijkstra(valves, total_flow, queue, minutes)
  end

  defp dijkstra(valves, total_flow, queue, minutes) do
    {step, queue} = Heap.split(queue)

    if step == nil do
      total_flow |> Enum.max_by(fn {path, flow} -> {flow, path} end)
    else
      {flow, minute, current, open} = step

      if(minute < minutes) do
        {rate, neighbors} = valves[current]

        {queue, total_flow} =
          if rate > 0 and not MapSet.member?(open, current) do
            new_flow = flow + rate * (minutes - minute - 1)
            queue = Heap.push(queue, {new_flow, minute + 1, current, MapSet.put(open, current)})

            total_flow =
              Map.put(total_flow, {current, minute + 1, MapSet.put(open, current)}, new_flow)

            {queue, total_flow}
          else
            {queue, total_flow}
          end

        {queue, total_flow} =
          for next <- neighbors,
              path = {next, minute + 1, open},
              Map.get(total_flow, path, -1) < flow,
              reduce: {queue, total_flow} do
            {queue, total_flow} ->
              queue = Heap.push(queue, {flow, minute + 1, next, open})
              total_flow = Map.put(total_flow, path, flow)
              {queue, total_flow}
          end

        dijkstra(valves, total_flow, queue, minutes)
      else
        dijkstra(valves, total_flow, queue, minutes)
      end
    end
  end

  def find_highest_flow_with_help(valves, start, minutes) do
    total_flow = %{{{start, start}, 0, 1, %MapSet{}} => 0}
    queue = Heap.max() |> Heap.push({0, 0, 1, {start, start}, %MapSet{}, {%MapSet{}, %MapSet{}}})
    dijkstra2(valves, total_flow, queue, minutes, 0)
  end

  defp dijkstra2(valves, total_flow, queue, minutes, max) do
    {step, queue} = Heap.split(queue)

    if step == nil do
      total_flow |> Enum.max_by(fn {path, flow} -> {flow, path} end)
    else
      {flow, minute, actor, {current1, current2}, open, {since_open1, since_open2}} = step

      if(minute < minutes) do
        if actor == 1 do
          {rate, neighbors} = valves[current1]

          {queue, total_flow} =
            if rate > 0 and not MapSet.member?(open, current1) do
              new_flow = flow + rate * (minutes - minute - 1)

              queue =
                Heap.push(
                  queue,
                  {new_flow, minute, 2, {current1, current2}, MapSet.put(open, current1),
                   {%MapSet{}, since_open2}}
                )

              total_flow =
                Map.put(
                  total_flow,
                  {{current1, current2}, minute, 2, MapSet.put(open, current1)},
                  new_flow
                )

              {queue, total_flow}
            else
              {queue, total_flow}
            end

          {queue, total_flow} =
            for next <- neighbors,
                path = {{next, current2}, minute, 2, open},
                Map.get(total_flow, path, -1) < flow,
                not MapSet.member?(since_open1, next),
                reduce: {queue, total_flow} do
              {queue, total_flow} ->
                queue =
                  Heap.push(
                    queue,
                    {flow, minute, 2, {next, current2}, open,
                     {MapSet.put(since_open1, next), since_open2}}
                  )

                total_flow = Map.put(total_flow, path, flow)
                {queue, total_flow}
            end

          dijkstra2(valves, total_flow, queue, minutes, max)
        else
          {rate, neighbors} = valves[current2]

          {queue, total_flow} =
            if rate > 0 and not MapSet.member?(open, current2) do
              new_flow = flow + rate * (minutes - minute - 1)

              queue =
                Heap.push(
                  queue,
                  {new_flow, minute + 1, 1, {current1, current2}, MapSet.put(open, current2),
                   {since_open1, %MapSet{}}}
                )

              total_flow =
                Map.put(
                  total_flow,
                  {{current1, current2}, minute + 1, 1, MapSet.put(open, current2)},
                  new_flow
                )

              {queue, total_flow}
            else
              {queue, total_flow}
            end

          {queue, total_flow} =
            for next <- neighbors,
                path = {{current1, next}, minute + 1, 1, open},
                Map.get(total_flow, path, -1) < flow,
                reduce: {queue, total_flow} do
              {queue, total_flow} ->
                queue =
                  Heap.push(
                    queue,
                    {flow, minute + 1, 1, {current1, next}, open,
                     {since_open1, MapSet.put(since_open2, next)}}
                  )

                total_flow = Map.put(total_flow, path, flow)
                {queue, total_flow}
            end

          dijkstra2(valves, total_flow, queue, minutes, max)
        end
      else
        max_flow = total_flow |> Map.values() |> Enum.max()
        if max_flow > max, do: IO.inspect({flow, max_flow})
        dijkstra2(valves, total_flow, queue, minutes, Enum.max([max, max_flow]))
      end
    end
  end

  def main(args) do
    Enum.map(args, fn path -> AOC.solve(path, &parse/1, &part1/1, &part2/1) end)
  end
end

File: python/src/2022/16_proboscidea_volcanium/aoc202216.py
"""AoC 16, 2022: Proboscidea Volcanium."""

# Standard library imports
import collections
import itertools
import pathlib
import sys

# Third party imports
import parse

PATTERNS = [
    parse.compile(
        "Valve {valve} has flow rate={rate:d}; tunnel leads to valve {valves}"
    ),
    parse.compile(
        "Valve {valve} has flow rate={rate:d}; tunnels lead to valves {valves}"
    ),
]


def parse_data(puzzle_input):
    """Parse input."""
    data = prune_graph(dict(parse_valve(line) for line in puzzle_input.split("\n")))
    return (
        {valve: tunnels for valve, (_, tunnels) in data.items()},
        {valve: flow_rate for valve, (flow_rate, _) in data.items()},
    )


def parse_valve(line):
    """Parse one line of input describing one valve.

    ## Examples:

    >>> parse_valve("Valve PT has flow rate=18; tunnel leads to valve GS")
    ('PT', (18, ['GS']))
    >>> parse_valve("Valve KE has flow rate=21; tunnels lead to valves ES, SQ, FS")
    ('KE', (21, ['ES', 'SQ', 'FS']))
    """
    if not any((match := pattern.parse(line)) for pattern in PATTERNS):
        raise ValueError(f"bad input: {line}")
    return match["valve"], (match["rate"], match["valves"].split(", "))


def prune_graph(valves):
    """Prune the graph to only keep the functioning valves.

    The AA valve is also kept since it's the starting point.

    ## Example:

    >>> valves = {"AA": (0, ["B", "C"]), "B": (3, ["AA", "D"]), "C": (5, ["AA"]), "D": (0, ["B", "E"]), "E": (7, ["D"])}
    >>> prune_graph(valves)
    {'AA': (0, {'B': 1, 'C': 1, 'E': 3}), 'B': (3, {'C': 2, 'E': 2}), 'C': (5, {'B': 2, 'E': 4}), 'E': (7, {'B': 2, 'C': 4})}
    """
    operational = [valve for valve, (rate, _) in valves.items() if rate > 0]
    paths = {
        node: bfs({valve: tunnels for valve, (_, tunnels) in valves.items()}, node)
        for node in valves
    }
    return {
        valve: (
            rate,
            {tunnel: paths[valve][tunnel] for tunnel in operational if tunnel != valve},
        )
        for valve, (rate, _) in valves.items()
        if rate != 0 or valve == "AA"
    }


def bfs(graph, start):
    """Find the shortest paths from the start node in a small graph."""
    queue = collections.deque([(0, start)])
    paths = {}
    while queue:
        distance, node = queue.popleft()
        if node in paths:
            continue
        paths[node] = distance
        for neighbor in graph[node]:
            queue.append((distance + 1, neighbor))
    return paths


def part1(data):
    """Solve part 1."""
    return best_flow(*data, 30)


def part2(data):
    """Solve part 2."""
    graph, flows = data

    return max(
        best_flow_subgraph(graph, flows, minutes=26, valves=me)
        + best_flow_subgraph(graph, flows, minutes=26, valves=elephant)
        for me, elephant in split_valves(graph.keys())
    )


def split_valves(valves):
    """Split responsibility for the valves in two.

    ## Example:

    >>> [(me, sorted(elephant)) for me, elephant in split_valves(["A", "B", "C"])]
    [({'A'}, ['B', 'C']), ({'B'}, ['A', 'C']), ({'C'}, ['A', 'B'])]
    """
    rooms = {valve for valve in valves if valve != "AA"}

    size = len(rooms) // 2
    for me in itertools.combinations(sorted(rooms), size):
        yield set(me), rooms - set(me)


def best_flow_subgraph(graph, flows, minutes, valves):
    """Calculate the best flow in a subgraph for the given valves."""
    subgraph = {
        v: {t: d for t, d in ts.items() if t in valves}
        for v, ts in graph.items()
        if v in valves or v == "AA"
    }
    subflows = {v: fr for v, fr in flows.items() if v in valves} | {"AA": 0}
    return best_flow(subgraph, subflows, minutes)


def best_flow(graph, flows, minutes):
    """Calculate the best flow that can be achieved in the given number of minutes."""
    max_flow = 0

    queue = collections.deque([(0, 0, "AA", set(graph) - {"AA"})])
    while queue:
        minute, flow, current, closed = queue.popleft()
        steps = graph[current]

        for step in closed:
            if steps[step] >= minutes - minute:
                continue
            new_flow = flow + flows[step] * (minutes - minute - steps[step] - 1)
            queue.append((minute + 1 + steps[step], new_flow, step, closed - {step}))
            if new_flow > max_flow:
                max_flow = new_flow
    return max_flow


def solve(puzzle_input):
    """Solve the puzzle for the given input."""
    data = parse_data(puzzle_input)
    yield part1(data)
    yield part2(data)


if __name__ == "__main__":
    for path in sys.argv[1:]:
        print(f"\n{path}:")
        solutions = solve(puzzle_input=pathlib.Path(path).read_text().rstrip())
        print("\n".join(str(solution) for solution in solutions))
